feat(networking): implement forward auth

ishanjain28 opened the pull request:
## Description

This PR adds the Forward Auth feature. It fixes the feedback received in previous attempt [here](https://github.com/fallenbagel/jellyseerr/pull/1341) 

1. A few different header names are allowed but they have to added to the allow list in the code. Supporting completely arbitrary headers was difficult.
2. Forward auth headers are only processed if request came from a trusted proxy and configuring trusted proxy is required to use the forward auth feature.
3. Forward auth can be done in 3 ways. 
    a. Verify the user field against `jellyfinUsername | plexUsername` columns. 
    b. Verify the email field against `email` column. 
    c. Verify both the user and email fields.

- Fixes #1221

## How Has This Been Tested?

This has been tested using the included test suite and manually in different environments. Manual test examples, 

1. With Seerr behind 1 trusted proxy and 2 trusted proxies.
2. With Seerr configured with no trusted proxies
3. Different header names in use and sometimes they were set, sometimes they were not set and I got the expected result every time.

## Screenshots / Logs (if applicable)
<img width="1595" alt="Screenshot 2025-04-06 at 17 42 14" src="https://github.com/user-attachments/assets/a2a3ac91-946f-4080-9fd2-e3c684e8df0b" />


## Checklist:

- [x] I have read and followed the contribution [guidelines](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md).
- [x] Disclosed any use of AI (see our [policy](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md#ai-assistance-notice))
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] Successful build `pnpm build`
- [x] Translation keys `pnpm i18n:extract`
- [ ] Database migration (if required)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Forward Authentication (SSO) via trusted proxies with configurable user/email headers and Advanced Network Settings UI (trusted IPv4/IPv6 lists, validation).

* **Documentation**
  * Added guide for configuring Forward Authentication with Traefik and example auth provider setup.

* **Chores**
  * Added runtime IP validation dependency.

* **Tests**
  * Updated end-to-end test covering trusted-proxy/forward-auth and restart modal behavior.

* **Quality**
  * Centralized server-side auth header handling and allowlist for forwarded headers.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ishanjain28 committed:
feat(forward-auth): addded support for forward auth

ishanjain28 committed:
feat(forward-auth): addded support for forward auth

ishanjain28 commented:
@fallenbagel @gauthier-th This PR is now ready for review!

fallenbagel commented on Dockerfile:
This file should not be changed by this PR

fallenbagel reviewed (changes requested):
I didn't fully review this yet but I noticed one issue. The Dockerfile. Also in terms of UI, I would recommend putting the forwardAuth stuff to be indented like how http(s) proxy is done:
![image](https://github.com/user-attachments/assets/d2a90d1c-659b-4c3e-a689-798ba280175d)

Also I would argue that this should be under the `advanced networking` section as this is not for the normal user. Wdyt about that @gauthier-th

gauthier-th commented:
> I didn't fully review this yet but I noticed one issue. The Dockerfile. Also in terms of UI, I would recommend putting the forwardAuth stuff to be indented like how http(s) proxy is done:
> ![image](https://github.com/user-attachments/assets/d2a90d1c-659b-4c3e-a689-798ba280175d)
> 
> Also I would argue that this should be under the `advanced networking` section as this is not for the normal user. Wdyt about that @gauthier-th 

100% agree

ishanjain28 commented:
@fallenbagel @gauthier-th Applied all the suggestions.
1. reverted changes to dockerfile
2. Moved the trusted proxies section, forward auth sections to advanced menu. forward auth depends on trusted proxy. 
3. Indented as requested 
<img width="1788" alt="Screenshot 2025-04-07 at 22 36 49" src="https://github.com/user-attachments/assets/8df67155-d213-4618-b828-6b7477455cbc" />

M0NsTeRRR commented on src/components/Settings/SettingsNetwork/index.tsx:
```diff
@@ -56,19 +67,85 @@ const SettingsNetwork = () => {
+          }
+          const addresses = value.split(',').map((value) => value.trim());
+          for (const address of addresses) {
+            if (address.indexOf('.') != -1) {
+              if (!Address4.isValid(address)) {
+                return ctx.createError({
+                  message: `Invalid IPv4 address: ${address}`,
+                });
+              }
+            } else if (address.indexOf(':') != -1) {
+              if (!Address6.isValid(address)) {
+                return ctx.createError({
+                  message: `Invalid IPv6 address: ${address}`,
+                });
+              }
```
```suggestion
            if (!Address4.isValid(address)) {
              return ctx.createError({
                message: `Invalid IPv4 address: ${address}`,
              });
            } else if (!Address6.isValid(address)) {
              return ctx.createError({
                message: `Invalid IPv6 address: ${address}`,
              });
            }
```

M0NsTeRRR replied:
Ah yes, sorry — I didn't read it carefully. At that point, we can have a single condition: if the IP is neither IPv4 nor IPv6, then we return an error.

ishanjain28 replied:
I wanted the message to more helpful and have the additional context. It should show if an ipv4 address is invalid or an ipv6 address is invalid along with the invalid address itself.

M0NsTeRRR replied:
I don't fully agree, as you're trying to guess user input based on `.` and `:`. If they were different inputs, I would agree with you, but no problem, still minor things. :smile:

M0NsTeRRR commented on src/utils/forwardAuthList.ts:
```diff
@@ -0,0 +1,8 @@
+  'Remote-Email',
+  'Remote-Name',
+  'Cf-Access-Authenticated-User-Email',
+  'X-authentik-username',
```
Use camel case like the other headers. It's still not mandatory, but let's stay consistent.

ishanjain28 replied:
These headers are defaults and come from Authentik, Authelia, Cloudflare.

M0NsTeRRR replied:
Yes, but headers are not case-sensitive. See [RFC 7230, Section 3.2](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2), and you're using them in a function with lowercase.

ishanjain28 replied:
That is very good point! Thank you. 

I'll convert them all to lower case since we are doing that anyway and make it a bit faster

M0NsTeRRR commented on server/middleware/auth.ts:
```diff
@@ -10,9 +10,29 @@ export const checkUser: Middleware = async (req, _res, next) => {
+    }
+  }
 
+  if (proxyIP && proxyIP.indexOf('.') != -1) {
```
Since we have `ip-address`, we can use `isValid` instead of `indexOf` to perform proper validation.

ishanjain28 replied:
We do not have to do validation here. The only way to reach this code is if the header was set by a trusted proxy and it is assumed to be correct. The only reason we do indexOf is to check if it's a v4 or a v6 address and then search the right store for it.

M0NsTeRRR commented on src/components/Settings/SettingsNetwork/index.tsx:
```diff
@@ -104,6 +186,19 @@ const SettingsNetwork = () => {
+              };
+              for (let value of values.trustedProxies.split(',')) {
+                value = value.trim();
+                if (value.indexOf('.') != -1) {
```
Since we have `ip-address`, we can use `isValid` instead of `indexOf` to perform proper validation.

ishanjain28 replied:
@M0NsTeRRR This is not doing validation. It is just for separating v4 and v6 addresses for pretty printing in the text field. We do not have to validate them.

If they are invalid, Right after they are rendered to the input, User will see an address validation error

ishanjain28 replied:
The ip-address library operates on strings/regex rather than constructing an int or [int;2] and have fast implementations for these operations. This will be hit on every request and I didn't want the needless validation to slow things down.

M0NsTeRRR replied:
Understable !

bash9810 commented:
excited for this, would love to setup authentik as the auth for jellyseerr, think this is the way to get that done.

ishanjain28 commented:
Rebased this PR on top of latest develop and fixed a small bug. Ready for review!

ishanjain28 commented:
Rebased the PR on top of latest develop and it is ready for review!

caesay commented:
Would really love to see this merged. This is probably the only thing preventing me from adopting Jellyseerr. I have a lot of users and could never justify them needing separate auth when I already have a totally unified auth setup with oidc at my reverse proxy.

The author seems to have done everything needed to get this going, and has now rebased this a couple of times. What do we need to do to get a review / merge from a maintainer?

WickedColdfront commented:
_This pull request has merge conflicts. Please resolve the conflicts so the PR can be successfully reviewed and merged._

Good lord.

fallenbagel commented:
> _This pull request has merge conflicts. Please resolve the conflicts so the PR can be successfully reviewed and merged._
> 
> Good lord. 

If you dont have anything productive to say in relation this PR, please refrain from commenting here. Merge conflicts are gonna occur now that we have restarted merging in new features after a feature freeze for the past couple of weeks in order to fix bugs. And this means reviewing open PRs and merging them **including this PR**.

M0NsTeRRR commented on server/middleware/auth.ts:
```diff
@@ -10,9 +10,32 @@ export const checkUser: Middleware = async (req, _res, next) => {
+  // The first header should be the client IP and the last header
+  // should be the address of proxy just upstream of us. We use that
+  // address to figure out if it should be trusted
+  let proxyIP = req.socket.remoteAddress;
+
+  const ipHeader = req.header('X-Forwarded-For');
+  if (ipHeader) {
+    const lastComma = ipHeader.lastIndexOf(',');
+    if (lastComma != -1) {
+      proxyIP = ipHeader.slice(lastComma + 1).trim();
+    }
+  }
```
I fully agree with @codyps and @0xSysR3ll :+1: and I'd go with the first option too.

M0NsTeRRR replied:
> reverse proxy are more than 1 l3 layer apart
I don't understand why it would not work.
Proxy <- (Network 1) -> Router <- (Network 2) -> Jellyseerr

In this case req.socket.remoteAddress will be equal to Proxy IP in the Network 1 it would not work only if you use NAT (edge case and I don't see why and where you will do that).

ishanjain28 replied:
This assumption does not hold true in more complex setups. I'll give two examples.

1. Running multiple instances of jellyseerr for HA and then using VIPs for routing traffic. The application will be in a different subnet and not see the ingress reverse proxy address in the socket address field. 

2. Applications deployed in k8s setups that use service mesh. Applications see address related to the mesh, not the reverse proxy. 

Sorry I am just repeating myself over and over.. 

Either you trust the proxy and expect it to do its job or you don't trust the proxy and don't use forward auth.

ishanjain28 replied:
1. Jellyseerr does not currently say that anywhere. I don't get how you arrived at that conclusion? The suggested alternative here will break those setups so the correct statment will be that, jellyseerr doesn't care if it breaks deployments like that. 
2. The linkerd/service mesh node talking to the service can change at any time. You would have to add the whole subnet in the trusted proxies list and then check if the socketaddress is in the same subnet. But, the attacker can just add a address from that subnet in the x-forwarded-for header, the check in jellyseerr will pass and you have the same problem again!  You will also have the same problem if attacker finds out your proxy address(that's talking to jellyseerr) and add that as a x-forwarded-for header + some fake values for the forward auth header. The check will once again pass and you have the same problem! 


To reiterate, there is no way to fix the reported problem in jellyseerr. You have no way to know what should be trusted and what shouldn't be trusted and you need to defer that job to the proxy. The proxy knows real client ip, it should be ignoring x-forwarded-for value in incoming request from a untrusted source and setting it.

Forward auth fundamentally relies on trusting the proxy. If you don't trust your proxy to do the right things, you obviously should not be using it. 


Do you see the problem yet?

M0NsTeRRR replied:
1. It have never been mentionned that it's supported and PR are not reviewed with this feature in mind. If people build infrastructure based on something that as never been supported even if it worked on some setup it can breaks yes but it could be a great enhancement for sure but it's a separate topic.
2.
> The linkerd/service mesh node talking to the service [...]

Thank you for detailing your example. It’s a complex setup you’re describing (and not used by a lot of people here) I haven’t used a service mesh in Kubernetes with Cilium or Istio for example, which is why I’m asking for details. Sorry if it bothers you; that’s not my intention.

> But, the attacker can just add a address from that subnet in the x-forwarded-for header, the check in jellyseerr will pass

To add to my previous answer (let me know if I’m wrong), but are you referring to the sidecar proxy inside the pod? If so, what I meant is that Jellyseerr should trust your Envoy proxy, and Envoy should be configured to trust only your service. This way, you establish a trusted chain.
If not can you point me to a schema or something like that, that describe the architecture you have in mind ?

Also, the code snippet we discus wouldn’t make sense to be set up [there](https://expressjs.com/en/guide/behind-proxies.html) as a custom function ?
 
> To reiterate, there is no way to fix the reported problem in jellyseerr. You have no way to know what should be trusted and what shouldn't be trusted and you need to defer that job to the proxy. The proxy knows real client ip, it should be ignoring x-forwarded-for value in incoming request from a untrusted source and setting it.

Yes I don't speak about that, I agree with you it's the meaning of trusting a proxy.

Janhouse replied:
Great PR @ishanjain28!
I concur that if you are using forward-auth, you likely already limit access to the service only through the proxy. It makes no sense to engineer something extremely complicated here since in a real-world scenario the trusted proxy is determined not just by and IP addresses, it frequently involves security headers and other mechanisms. And it is not the job of the service to do those checks. Forward auth is supposed to be a simple way to authenticate using external systems.
Just add a message to the user that if enabling it, user should allow access only through the proxy.

kevincox replied:
The current code seems wrong. It is extracting the last element of the `X-Forwarded-For` which would be the client IP in most setups, so why is it being called `proxyIP`? In multi-proxy setups this would be the second-last proxy IP but that also seems like not the value that we want to check?

IMHO there are two reasonable options here:

1. Use a different header and anyone who enables this must set it with the client IP and prevent access from any other sources.
2. Walk backwards from the socket IP then right-to-left via `X-Forwarded-For` the client IP is then the first non-trusted IP.

The current code is doing something else that I don't understand.

ishanjain28 replied:
@kevincox The `X-Forwarded-For` RFC specifies this order.

> client, proxy1, proxy2

where proxy2 is closer to the service. See the [wiki page](https://en.wikipedia.org/wiki/X-Forwarded-For)

Historically, there has been some confusion and http servers implementing this incorrectly. New ones like Caddy do it correctly :) 

The current code expects `X-Forwarded-For` to have a value like, 
```
X-Forwarded-For: 10.0.10.5, 10.0.50.3
```
OR
```
X-Forwarded-For: 10.0.50.3
```

In first case, it'll read the address of closest proxy(10.0.50.3), in second case it'll read the whole value after trimming it. That can happen if the proxy itself is sending requests(maybe for active healthchecks?) although it is very very unlikely to hit that path.

kevincox replied:
I am well aware how `X-Forwarded-For` works.

For a lot of people the setup will be a single proxy. So the `X-Forwarded-For` will just contain the client IP. This code will then call that the `proxyIP` then (presumably) fail the trusted subnet check. What is actually intended in this case is to check that `req.socket.remoteAddress` is a trusted IP.

For the example of `X-Forwarded-For: 10.0.10.5, 10.0.50.3` it seems like a very strange case:

1. You ignore the trustworthiest of the nearest proxy.
2. You check the trustworthiness of the second proxy (`10.0.50.3`)

I guess it could be reasonable if you implicitly trust the first proxy (for example you are listening on localhost and trust anything that can connect to localhost) but this is a weird setup, and it would still be better to actually check the trustworthiness of the nearest proxy.

Honestly this really raises the question what is the point of this check at all. It seems to be hardcoded to check the IP of one hop away which seems like a somewhat niche use case (maybe if you are using a global CDN then a local reverse proxy). But I doubt most people care at all about that second hop. If the nearest hop is implicitly trusted it may as well do all checks and clear undesirable headers.

ishanjain28 replied:
> For a lot of people the setup will be a single proxy. So the X-Forwarded-For will just contain the client IP. 

Wrong. See the documentation, please. In this situation, it'll still have two address. The client ip followed by the proxy's address.

> I guess it could be reasonable if you implicitly trust the first proxy (for example you are listening on localhost and trust anything that can connect to localhost) but this is a weird setup, and it would still be better to actually check the trustworthiness of the nearest proxy.

This is incorrect. The trustworthiness of a proxy in the chain is determined by the proxy downstream to it. i.e. if 10.0.10.5 can be trusted will be determined by 10.0.50.3. The proxy at 10.0.50.3 will discard `X-Forwarded-For` header completely and set it to `{socketRemoteAddress} {proxy's own address}` and if the request came from a trusted source, it'll append it's own address and trust the rest of it is correct. 

The service itself(seerr) can not make the decision if it should trust all the proxies in the chain. The proxies in the chain have to calculate that for their source. 

> Honestly this really raises the question what is the point of this check at all. It seems to be hardcoded to check the IP of one hop away which seems like a somewhat niche use case (maybe if you are using a global CDN then a local reverse proxy). But I doubt most people care at all about that second hop. If the nearest hop is implicitly trusted it may as well do all checks and clear undesirable headers.

Please see existing implementations of how this is done in applications and proxies and read documentation on the header itself. You seem to have a misunderstanding here.

kevincox replied:
From your own source:

> The general format of the field is:
>
> ```
> X-Forwarded-For: client, proxy1, proxy2
> ```
>
> where the value is a comma+space separated list of IP addresses, the left-most being the original client, and each successive proxy that passed the request adding the IP address where it received the request from. In this example, the request passed through proxy1, proxy2, and **then proxy3 (not shown in the header). proxy3 appears as remote address of the request**.
>
> https://en.wikipedia.org/wiki/X-Forwarded-For#Format

Emphasis mine.

ishanjain28 replied:
@kevincox Actually you are righgt about this! 

So it needs to look just the socket address and check if it is _trusted_.

kevincox replied:
> So it needs to look the last address in x-forwarded-for

I'm pretty sure that we don't need to look at `X-Forwarded-For` at all. We only need to do that if we want to find the client IP (which I don't think Seerr currently does).

What I would recommend:
1. Check the socket address.
2. Warn users that whatever proxies they mark as trusted are responsible for setting/clearing all of the relevant headers.

ishanjain28 replied:
> I'm pretty sure that we don't need to look at X-Forwarded-For at all.

Yes, agreed. It took a second for my comment edit to go through. We already have a message to only enable this behind trusted proxies. I'll make the other change.

ishanjain28 replied:
Thanks again @kevincox for pointing this out. I have fixed this now and also tested it in my setup with 1 proxy and 2 proxies. it was previously working fine before because traffic from public internet goes through two proxies and I had added addresses of both proxies in trusted proxies list in seerr and I never tested it at home where traffic would go through a single proxy.

brodock commented on server/lib/settings/index.ts:
```diff
@@ -547,6 +560,15 @@ class Settings {
+        trustedProxies: {
+          v4: [],
+          v6: [],
+        },
```
if forward auth is going to take a while, can we have the trusted proxies list merged in separate? as it is currently, enabling "trustProxy" has security implications as requests can be forged, unless a firewall prevents accessing the instance directly into the 5055 port.

ishanjain28 commented on server/middleware/auth.ts:
```diff
@@ -21,9 +44,68 @@ export const checkUser: Middleware = async (req, _res, next) => {
+      user = await userRepository.findOne({
+        where: query,
+      });
+    }
```
This is fine!

ishanjain28 commented on src/i18n/locale/en.json:
```diff
@@ -1013,41 +1013,6 @@
-  "components.Settings.SettingsNetwork.validationApiRequestTimeout": "You must provide a valid timeout value",
-  "components.Settings.SettingsNetwork.validationDnsCacheMaxTtl": "You must provide a valid maximum TTL",
-  "components.Settings.SettingsNetwork.validationDnsCacheMinTtl": "You must provide a valid minimum TTL",
-  "components.Settings.SettingsNetwork.validationProxyPort": "You must provide a valid port",
```
I am not sure why is it removing this ? I ran pnpm run i18n:extract a few times and it seems to be removing this block every time.

ishanjain28 replied:
Fixed

kevincox commented on src/utils/forwardAuthList.ts:
```diff
@@ -0,0 +1,13 @@
+  'remote-name',
+  'cf-access-authenticated-user-email',
+  'x-authentik-username',
+  'x-authentik-email',
```
Why is this hardcoded? Shouldn't this be based on the configured auth headers?

ishanjain28 replied:
@kevincox The list is presented in the forward auth UI in network settings(See screenshot in the first message) and It covers forward auth headers used by popular identity providers. I preferred this to having users type it manually because I felt this was easier. 

We'll have two options if some new IdP comes out or if people need to enter some other header name. 
1. They can edit the JSON file directly. 
2. We can add that header name in Seerr. 

Usually, the header names are kind of standardized because applications need some consistency and not every self-hostable app offers an option to edit the forward auth header.

kevincox replied:
Why not just use this list as a autocomplete and let users use whatever value they need? I see no point restricting it.

> They can edit the JSON file directly.

Can they? This constant is used in the code and I don't see any way for configuration to override it.

https://github.com/seerr-team/seerr/pull/1564/changes/BASE..902c39b47b3fa78fc0ab51656db00e4c9155bf3b#diff-8e0c630fb9bd5297c001b08a3d6810a6cab9a9a05e8745d7c2151ccaecb56093R19

> Usually, the header names are kind of standardized because applications need some consistency and not every self-hostable app offers an option to edit the forward auth header.

In my experience almost all allow editing the header.

ishanjain28 replied:
I personally do not foresee this as something people may need(specially, since there is an option to edit the config file directly). However, If the maintainers _want_ this now, I'll look into adding it. Otherwise, it can be added in future if/when someone asks for it.

M0NsTeRRR replied:
Providing a default list is a good approach, but it’s also important to allow users to specify their own values. This way, we avoid having to review pull requests just to add missing headers, which can delay users who need additional options while waiting for their PR to be merged.

We also want to provide the best user experience, so we should avoid asking users to edit this file for common operations.

So I agree with @kevincox approach

> list as a autocomplete and let users use whatever value they need

ishanjain28 replied:
@M0NsTeRRR Okay, that makes sense! My preference is to keep the PR in the current state and eventually if someone needs it, it can be changed to a auto complete field. How does that sound ? The current list covers forward auth headers used by cloudflare, authentik, authelia and iirc 1 or 2 more IdP

M0NsTeRRR replied:
I would prefer to have it day one 😄

wdyt @seerr-team/seerr-core ?

ishanjain28 replied:
Sure but I don't really want to go make more changes in that part of the code again :/

fallenbagel replied:
> I would prefer to have it day one 😄
> 
> wdyt [@seerr-team/seerr-core](https://github.com/orgs/seerr-team/teams/seerr-core) ?

I agree with the autocomplete approach. The hardcoded list also has a practical issue I believe (unless I'm mistaken?). The `getAuthHeaders` in `serverSidePropsHelpers` only forwards headers in this allowlist during SSR, so any custom header set via `settings.json` won't be forwarded on SSR requests right? Wouldn't that break forward auth on server-rendered pages. So imo, the autocomplete with the current list as suggestions but allowing custom values would fix both the UX and this issue.

Janhouse replied:
Also having a list of headers looks like a security vulnerability. Any client can pass any header they like. Unless proxy is set up in a way where it filters out all of these, they will be passed to seer from client.
I think you are putting in too much effort to make this feature "beginner friendly". Forward auth will never be a one click setup, so it makes no sense to try to make it like that on seerr side.

ishanjain28 replied:
> The getAuthHeaders in serverSidePropsHelpers only forwards headers in this allowlist during SSR, so any custom header set via settings.json won't be forwarded on SSR requests right? Wouldn't that break forward auth on server-rendered pages.

Yess, this is correct. I also vaguely remember this was the reason I went with a allow list instead of allowing anything. The header needs to be copied in `getAuthHeaders`. I am not sure if I can get the settings object there and read the configured header names ? 
I'll check this again

> Also having a list of headers looks like a security vulnerability. Any client can pass any header they like. Unless proxy is set up in a way where it filters out all of these, they will be passed to seer from client.

The config needs to be aligned in the proxy and the application. If the proxy sets `Remote-User` but application is looking at `X-Authentik-User`, then that's a incorrectly configured. That'll also be a problem with completely custom header names.

M0NsTeRRR commented on docs/extending-seerr/forward-auth.md:
```diff
@@ -0,0 +1,77 @@
+If the user has no email set, the username will also work
+:::
+
+## Example with Goauthentik and Traefik
```
Regarding the authentik documentation, I think it would be more effective to update the existing authentik integration guide available [here](https://github.com/goauthentik/authentik/blob/main/website/integrations/media/jellyseerr/index.md) and add a link here instead, wdyt ?

ishanjain28 replied:
@M0NsTeRRR Makes sense! Is it okay if that goes in a follow up PR once update is merged in Authentik ?

M0NsTeRRR replied:
Yes sure

ishanjain28 replied:
@M0NsTeRRR Or honestly, I'll prefer to skip it completely. I don't use traefik and I don't have a setup to test the config and make sure it is correct. That part of the document was carried over from previous 1-2 people who worked on this feature.

M0NsTeRRR replied:
> I'll prefer to skip it completely. I don't use traefik and I don't have a setup to test the config and make sure it is correct.

Are you referring to Authentik as well, or just Traefik doc ?

If you can’t test it, we can remove it unless another maintainer can step in. I think @gauthier-th or @0xSysR3ll use Docker setups, so they might have Traefik available.

ishanjain28 replied:
> Are you referring to Authentik as well, or just Traefik doc ?

Both. My setup is Authelia / Caddy / LXC Containers so I can't test it. That document was carried over from previous contributors.

M0NsTeRRR replied:
Okay let's see if others maintainers can step in for this part and if not we will remove it.

Would you mind updating this for authelia and link it here as you have a working setup ?

https://www.authelia.com/integration/openid-connect/clients/jellyseerr/
https://github.com/authelia/authelia/blob/master/docs/content//integration/openid-connect/clients/jellyseerr/index.md

ishanjain28 replied:
Added docs in trusted header sso section here https://github.com/authelia/authelia/pull/11757

M0NsTeRRR replied:
I’ll let you add a link for the Authelia configuration.

You can remove the Authentik documentation for now we’ll add it back once it’s updated for Seerr.

For the Traefik part, I’ll wait for input from the other maintainers before making any changes.

M0NsTeRRR commented on package.json:
```diff
@@ -70,6 +70,7 @@
     "http-proxy-agent": "^7.0.2",
     "https-proxy-agent": "^7.0.6",
     "lodash": "4.17.23",
+    "ip-address": "^10.0.1",
```
You can bump to 10.1.0

ishanjain28 committed:
fix: read last proxy address from remote socket address

ishanjain28 committed:
fix: read last proxy address from remote socket address

ishanjain28 committed:
chore: upgrade ip-address dep

ishanjain28 committed:
chore: upgrade ip-address dep

fallenbagel commented on server/middleware/auth.ts:
```diff
@@ -22,13 +37,63 @@ export const checkUser: Middleware = async (req, _res, next) => {
       where: { id: req.session.userId },
     });
-  }
+  } else if (
+    settings.network.trustProxy &&
+    settings.network.forwardAuth.enabled &&
+    trustedProxy
+  ) {
```
There's no validation here to prevent `forwardAuth.enabled = true` with an empty `trustedProxies` list. Wouldn't this be confusing UX since in that case `trustedProxy` will always be false so forward auth will silently do nothing?

ishanjain28 replied:
@fallenbagel `trustedProxy` can still be true if the request came from `localhost`. `localhost` is implicitly always trusted.

fallenbagel commented on server/middleware/auth.ts:
nit: I would prefer strict equality is used for consistency with the rest of the codebase instead of `==` / `!=`

fallenbagel commented on server/middleware/auth.ts:
```diff
@@ -10,9 +10,24 @@ export const checkUser: Middleware = async (req, _res, next) => {
+    trustedProxy =
+      socketAddress == '::1' ||
+      settings.network.trustedProxies.v6.includes(socketAddress);
+  }
```
`indexOf('.')` for IP version detection will misclassify IPv4-mapped IPv6 addresses like `::ffff:127.0.0.1` as they contain both `.` and `:`, so they'll hit the v4 branch but won't match anything in the v4 trusted list. I would suggest using `net.isIPv4()` / `net.isIPv6()` from Node's `net` module instead as that handles mapped addresses correctly. This would be consistent with our codebase too as we use that in:
- https://github.com/seerr-team/seerr/blob/061121c326d416b6d6c126544a65bc2bb848e51f/server/routes/auth.ts#L296-L300
- https://github.com/seerr-team/seerr/blob/061121c326d416b6d6c126544a65bc2bb848e51f/server/routes/user/usersettings.ts#L407-L411

fallenbagel commented on docs/extending-seerr/forward-auth.md:
```diff
@@ -0,0 +1,77 @@
+# Forward auth
+
+You can use [forward-auth](https://doc.traefik.io/traefik/middlewares/http/forwardauth/) mechanism to log into Jellyseer.
```
```suggestion
You can use [forward-auth](https://doc.traefik.io/traefik/middlewares/http/forwardauth/) mechanism to log into Seerr.
```

fallenbagel commented on seerr-api.yml:
```diff
@@ -9,6 +9,9 @@ info:
     - **Cookie Authentication**: A valid sign-in to the `/auth/plex` or `/auth/local` will generate a valid authentication cookie.
     - **API Key Authentication**: Sign-in is also possible by passing an `X-Api-Key` header along with a valid API Key generated by Seerr.
+
+    Additionally, if forward authentication is enabled, Jellyseerr will authenticate as the user set in the header defined by `forwardAuth.userHeader`.
```
```suggestion
    Additionally, if forward authentication is enabled, Seerr will authenticate as the user set in the header defined by `forwardAuth.userHeader`.
```

ishanjain28 committed:
fix: address PR feedback, stricter comparisons, corrected typos, use net lib for ip-addr

ishanjain28 committed:
fix: address PR feedback, stricter comparisons, corrected typos, use net lib for ip-addr

gauthier-th commented on server/routes/settings/index.ts:
```diff
@@ -91,6 +91,7 @@ settingsRoutes.get('/network', (req, res) => {
 settingsRoutes.post('/network', async (req, res) => {
   const settings = getSettings();
 
+  settings.network.trustedProxies = { v4: [], v6: [] };
```
Why would you reset that?

ishanjain28 replied:
@gauthier-th I don't remember. I don't see any reason why we should do this so I'll revert this and test.

ishanjain28 replied:
@gauthier-th This is required to unset proxies.

The web interface gets all proxies and then sends a response with updated proxies. We clear out existing proxies in the config and then set it to the proxies in received response. 

If we don't clear it, it'll merge the existing set of proxies with the received set and it'll never actually remove the proxy that was removed by the user.

gauthier-th replied:
Got it. Could you please add a brief comment to clarify this? It’s not immediately clear at first glance.

ishanjain28 committed:
fix: switch from hardcoded list of headers to allowing any with autocomplete

ishanjain28 committed:
fix: switch from hardcoded list of headers to allowing any with autocomplete

fallenbagel reviewed (changes requested):
@ishanjain28 Sorry for the delayed review on this. Back when this PR was originally opened, we were a small team juggling a lot of PRs and were in the middle of a significant merge (including early discussions with the Overseerr team), so things slipped through the cracks.

> @fallenbagel @gauthier-th Applied all the suggestions.
> 
>     1. reverted changes to dockerfile
> 
>     2. Moved the trusted proxies section, forward auth sections to advanced menu. forward auth depends on trusted proxy.
> 
>     3. Indented as requested
> 
> 
> <img alt="Screenshot 2025-04-07 at 22 36 49" width="1788" src="https://private-user-images.githubusercontent.com/7921368/431039884-8df67155-d213-4618-b828-6b7477455cbc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYyNjYwOTMsIm5iZiI6MTc3NjI2NTc5MywicGF0aCI6Ii83OTIxMzY4LzQzMTAzOTg4NC04ZGY2NzE1NS1kMjEzLTQ2MTgtYjgyOC02Yjc0Nzc0NTVjYmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTVUMTUwOTUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDZkOTA3ZWQ3OWE1MzI3OTc1MTQ3ZjkzZmJhYTFhMWE5YWYzYmEzNjQ3MjNkNjczNDllYWNhYzIwZjNmOTA4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.NPNBj3c3Gir64glcz5nMFOJskNaZbeUlhZWLPBPijWk">

ref: https://github.com/seerr-team/seerr/pull/1564#issuecomment-2784034354

We're now asking you to partially revert the above. Here's the background:

- The advanced networking menu was removed in https://github.com/seerr-team/seerr/commit/a488f850f3d6f0d9896362db873bd86ec1b7040a#diff-fe3711be57274a5af941ff098722503db9e681aae8db4c0377b110cec36cae1cL384-L423 after we dropped the DNS server setting (#1416) and `forceIpv4First` at that point, there was nothing left to put in an "advanced" section.
- `forceIpv4First` was later re-added, but we decided to keep the networking tab flat and simple rather than reintroducing the advanced section (#1719).
- That decision was made without circling back to this PR as we did not have time to review this PR and since then it was forgotten. Sorry for our oversight, and for the confusion.

Could you restructure the proxy settings back to the following layout?
```
[X] Enable Proxy Support
    └── Trusted Proxies
    └── Forward Authentication
          └── User Header
          └── Email Header
```

Appreciate your patience with the back-and-forth, and sorry for the extra work!

ishanjain28 committed:
fix: address PR feedback

1. Add comment explaining the trustedProxies setter in server.
2. Reformat header autocompletes and use toLowerCase when checking requests.
3. Remove Advanced Network Settings and restructure settings page.

ishanjain28 committed:
fix: address PR feedback

1. Add comment explaining the trustedProxies setter in server.
2. Reformat header autocompletes and use toLowerCase when checking requests.
3. Remove Advanced Network Settings and restructure settings page.

Xatrekak committed:
Merge pull request #1 from Xatrekak/add-auto-provisioning-gui

Add auto provisioning gui

ishanjain28 committed:
fix(settingsnetwork): ran i18n:extract, reformatted code and fixed a build failure

ishanjain28 committed:
fix(settingsnetwork): ran i18n:extract

Xatrekak commented on server/middleware/auth.ts:
```diff
@@ -22,13 +42,120 @@
+        });
+        await userRepository.save(user);
+        logger.info(
+          `Auto-provisioned user via forward-auth: ${provisionUsername}`,
```
provisionUsername is a derived variable from the header value of an already authorized user from an upstream identity provider through a trusted proxy. The code path is only reachable through the forward authentication mechanism and not the typical username/password workflow. This field is not a user-provided value in the sense that an arbitrary value could be passed.

The only intended possible values are the user portion of an email address or the username provided by a remote-user header and in both cases must have already been approved by the admin in the upstream identity provider.

Xatrekak commented on server/middleware/auth.ts:
```diff
@@ -22,13 +42,120 @@
+        );
+      } catch (e) {
+        logger.error(
+          `Failed to auto-provision forward-auth user ${provisionUsername}`,
```
See previous comment https://github.com/seerr-team/seerr/pull/1564#discussion_r3420044343
